### PR TITLE
Add teardown method to BaseProfiler.

### PR DESCRIPTION
**Pull request by camruta**
## What does this PR do?

This PR adds a teardown method to BaseProfiler class. With the method, the different instances can handle arbitrary post-profiling steps outside of __del__, such as closing output file how they want to, or do the log stitching.

This was brought up in PR 6067 review here:
https://github.com/PyTorchLightning/pytorch-lightning/pull/6067/files/928e680a24a0483bff3ebef100c531adc09acce3?file-filters%5B%5D=.py#r583854357

### CODE REVIEW

**Review by ananthsub — COMMENTED** [review]
thanks for working on this @camruta ! A few things we should add:
- Where in the trainer do you want to call `teardown` 
- Unit tests to exercise that teardown works as expected for each of the provided profilers

**Comment by camruta** [comment]
> thanks for working on this @camruta ! A few things we should add:
> 
> * Where in the trainer do you want to call `teardown`

As suggested by you in the offline discussion, we will call this method in trainer before calling the call_teardown_hook. I'll update the code.

> * Unit tests to exercise that teardown works as expected for each of the provided profilers

Adding unit tests.

**Review by tchaton — COMMENTED** [review]
LGTM ! Please, add a test to make sure the function is being called.

**Comment by camruta** [comment]
> LGTM ! Please, add a test to make sure the function is being called.

Thanks for the review! Added the unite test.

**Comment by camruta** [comment]
> thanks for working on this @camruta ! A few things we should add:
> 
> * Where in the trainer do you want to call `teardown`
Added profiler.teardown call in trainer.py.
> * Unit tests to exercise that teardown works as expected for each of the provided profilers
Added unite test.

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
...
 
+def test_profiler_teardown(tmpdir):
+    """
+    This test checks if profiler teardown method is called when
+    trainer is exiting.
+    """
+    profilerSimple = SimpleProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
+    profilerAdvanced = AdvancedProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
+    profilerPytorch = PyTorchProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"), local_rank=0)
+
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:1615]:
```suggestion
    simple_profiler = SimpleProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
    advanced_profiler = AdvancedProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"))
    pytorch_profiler = PyTorchProfiler(output_filename=os.path.join(tmpdir, "profiler.txt"), local_rank=0)

```

File: tests/trainer/test_trainer.py [hunk 2]
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
...
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        profiler=profilerSimple,
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:1620]:
```suggestion
        profiler=simple_profiler,
```

File: tests/trainer/test_trainer.py [hunk 3]
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
...
+    )
+    trainer.fit(model)
+
+    assert profilerSimple.output_file.closed
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:1624]:
```suggestion
    assert simple_profiler.output_file.closed
```

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
...
+
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        profiler=profilerAdvanced,
+    )
+    trainer.fit(model)
+
+    assert profilerAdvanced.output_file.closed
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:1634]:
```suggestion
    assert advanced_profiler.output_file.closed
```

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
...
+
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        fast_dev_run=True,
+        profiler=profilerPytorch,
+    )
+    trainer.fit(model)
+
+    assert profilerPytorch.output_file.closed
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:1644]:
```suggestion
    assert pytorch_profiler.output_file.closed
```

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -1604,6 +1604,46 @@ def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
     trainer.fit(model)
 
 
+def test_profiler_teardown(tmpdir):
```
Comment (ananthsub) [inline tests/trainer/test_trainer.py:1607]: you could add a dummy profiler which sets profiler state in `teardown` which you can use to verify in the unit test that `teardown` is called

do we have a separate `test_profilers.py` ? i think it'd be useful to split out from the trainer. what do you think @tchaton ?

**Review by ananthsub — COMMENTED** [review]
mostly style nits and a comment on unit testing profilers

File: tests/trainer/test_trainer.py (hunk 6 above)
Comment (carmocca) [reply tests/trainer/test_trainer.py:1607]: > do we have a separate test_profilers.py ? i think it'd be useful to split out from the trainer

Yes! To `tests/test_profiler.py`

I have actually moved them in my (very) WIP PR to refactor profilers

Comment (carmocca) [reply tests/trainer/test_trainer.py:1607]: Moved in https://github.com/PyTorchLightning/pytorch-lightning/commit/51c9260fad5b1ed3b4e41a9ebf460bf4c609fe2a

**Review by carmocca — APPROVED** [review]
Merged master and applied some fixes! LGTM

File: CHANGELOG.md [hunk 7]
```diff
@@ -37,6 +39,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added arg to `self.log` that enables users to give custom names when dealing with multiple dataloaders ([#6274](https://github.com/PyTorchLightning/pytorch-lightning/pull/6274))
 
 
+- Added `teardown` method to `BaseProfiler` to enable subclasses defe post-profiling steps outside of `__del__` ([#6370](https://github.com/PyTorchLightning/pytorch-lightning/pull/6370))
```
Comment (ananthsub) [inline CHANGELOG.md:42]:
```suggestion
- Added `teardown` method to `BaseProfiler` to enable subclasses to do post-profiling steps outside of `__del__` ([#6370](https://github.com/PyTorchLightning/pytorch-lightning/pull/6370))
```

**Review by ananthsub — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**